include/cantera/kinetics/ReactionRate.h

```diff
@@ -11,473 +11,180 @@
-    virtual ~ReactionRateBase() {}
+    ReactionRate() : m_rate_index(npos) {}
+
+    // Copy constructor and assignment operator need to be defined because of the
+    // #m_evaluator member that can't (and shouldn't) be copied.
+    ReactionRate(const ReactionRate& other)
+        : m_input(other.m_input)
+        , m_rate_index(other.m_rate_index)
+    {}
+
+    ReactionRate& operator=(const ReactionRate& other) {
+        if (this == &other) {
+            return *this;
+        }
+        m_input = other.m_input;
+        m_rate_index = other.m_rate_index;
+        return *this;
+    }
+
+    virtual ~ReactionRate() = default;
+
+    //! Create a rate evaluator for reactions of a particular derived type.
+    //! Derived classes usually implement this as:
+    //!
+    //! ```.cpp
+    //! unique_ptr<MultiRateBase> newMultiRate() const override {
+    //!     return unique_ptr<MultiRateBase>(new MultiBulkRate<RateType, DataType>);
+    //! ```
+    //!
+    //! where `RateType` is the derived class name and `DataType` is the corresponding
+    //! container for parameters needed to evaluate reactions of that type.
+    virtual unique_ptr<MultiRateBase> newMultiRate() const = 0;
 
-public:
-    //! Identifier of reaction type
-    virtual std::string type() const = 0;
+    virtual const std::string type() const = 0;
 
-    //! Create multi-rate evaluator
-    virtual unique_ptr<MultiRateBase> newMultiRate() const = 0;
+    //! Set parameters
+    //! @param node  AnyMap object containing reaction rate specification
+    //! @param rate_units  unit definitions specific to rate information
+    virtual void setParameters(const AnyMap& node, const UnitStack& units) {
+        m_input = node;
+    }
 
-    //! Update reaction rate data based on temperature
-    //! @param T  temperature [K]
-    virtual void update(double T) = 0;
+    //! Return the parameters such that an identical Reaction could be reconstructed
+    //! using the newReaction() function. Behavior specific to derived classes is
+    //! handled by the getParameters() method.
+    AnyMap parameters() const {
+        AnyMap out;
+        getParameters(out);
+        return out;
+    }
 
-    //! Update reaction rate data based on temperature and pressure
-    //! @param T  temperature [K]
-    //! @param P  pressure [Pa]
-    //! @param concm  third-body concentration (if applicable)
-    virtual void update(double T, double P, double concm=0.) = 0;
+    //! Check basic syntax and settings of reaction rate expression
+    virtual void check(const std::string& equation, const AnyMap& node) {}
 
-    //! Update reaction rate data based on bulk phase
-    //! @param bulk  object representing bulk phase
-    //! @param concm  third-body concentration (if applicable)
-    virtual void update(const ThermoPhase& bulk, double concm=0.) = 0;
+    //! Validate the reaction rate expression
+    virtual void validate(const std::string& equation, const Kinetics* kin = 0) {}
```

Comment:
I think it might be better to have the function signature for this be `void validate(const std::string& equation, const Kinetics& kin)`. With the exception of `Plog::validate(string&)`, this function is new for all rate types anyway.

Reply:
Ok. This requires some transitional handling (need to keep the old signature for `PlogRate::validate`), but that's minor.